clickhouse: add cookbook recipe on how to integrate postgresql with terraform

wojcik-dorota opened the pull request:
https://aiven.atlassian.net/browse/DOC-99

wojcik-dorota commented:
@dewan-ahmed , this is now ready for review. Please have a look - thank you

aris-aiven commented on docs/tools/terraform/reference/cookbook/postgres-clickhouse-integration-recipe.rst:
```diff
@@ -0,0 +1,160 @@
+  resource "aiven_pg_database" "mydatabase" {
+    project       = var.project_name
+    service_name  = aiven_pg.postgres.service_name
+    database_name = "sensors"
```
Change the database name to something more specific, precising that it's comprised of dimension tables
```suggestion
    database_name = "sensor_dims"
```

aris-aiven commented on docs/tools/terraform/reference/cookbook/postgres-clickhouse-integration-recipe.rst:
```diff
@@ -0,0 +1,160 @@
+  }
+
+  // Postgres sensors database
+  resource "aiven_pg_database" "mydatabase" {
```
Change the resource name to something more specific as well
```suggestion
  resource "aiven_pg_database" "sensor_dims" {
```

aris-aiven commented on docs/tools/terraform/reference/cookbook/postgres-clickhouse-integration-recipe.rst:
```diff
@@ -0,0 +1,160 @@
+    project                  = var.project_name
+    integration_type         = "clickhouse_postgresql"
+    source_service_name      = aiven_pg.postgres.service_name
+    destination_service_name = aiven_clickhouse.clickhouse.service_name
```
Here we need to add a configuration block that has been surfaced by the latest terraform provider, for instance

```suggestion
    destination_service_name = aiven_clickhouse.clickhouse.service_name
    clickhouse_postgresql_user_config {
      databases {
        database = aiven_pg_database. sensor_dims.database_name
        schema = "public"
      }
  }
```

aris-aiven commented on docs/tools/terraform/reference/cookbook/postgres-clickhouse-integration-recipe.rst:
```diff
@@ -0,0 +1,160 @@
+* Resource ``aiven_clickhouse`` creates an Aiven for ClickHouse service with the project name, the cloud name (provider, region, zone), the Aiven service plan, and the service name as specified in the ``services.tf`` file.
+* ``aiven_clickhouse_database`` resource creates a database that can be used to store high-throughput measurement data and to create new tables and views to process this data.
+* ``aiven_pg`` resource creates an Aiven for PostgreSQL service.
+* * ``aiven_pg_database`` resource creates the ``sensors`` database.
```
Correct the database name
```suggestion
* * ``aiven_pg_database`` resource creates the ``sensor_dims`` database.
```

aris-aiven commented on docs/tools/terraform/reference/cookbook/postgres-clickhouse-integration-recipe.rst:
```diff
@@ -0,0 +1,160 @@
+* ``aiven_clickhouse_database`` resource creates a database that can be used to store high-throughput measurement data and to create new tables and views to process this data.
+* ``aiven_pg`` resource creates an Aiven for PostgreSQL service.
+* * ``aiven_pg_database`` resource creates the ``sensors`` database.
+* ``aiven_service_integration`` resource creates the integration between the Aiven for PostgreSQL and Aiven for ClickHouse services.
```
This will result in the creation of a ``service_postgres-gcp-us_sensor_dims_public`` database in ClickHouse allowing the users to access the ``sensor_dims`` database for the ``postgres-gcp-us`` service.
It might be worth to point it out as the name might not be self-evident.

aris-aiven reviewed (dismissed):
Some comments around resource naming and the recent changes introduced by the terraform provider.

dewan-ahmed commented:
I'll review this on Monday.

dewan-ahmed commented:
https://github.com/aiven/devportal/pull/1591/commits/69f244fd5a4515df5d05040dc11479017f38e939 resolves the _toc.yml conflict.

dewan-ahmed reviewed (commented):
Hi, here are the review suggestions:

1. Please use `version = "~> 3.9.0"` instead of `version = ">= 3.7"`
2. "This article shows by way of example..." phrase has been used already in the two other ClickHouse recipes. To avoid monotonous reading, please rephrase this sentence.
3. I missed to note in the previous two ClickHouse recipes but all of our cookbook recipes include Mermaid diagrams. Kindly include Mermaid diagrams for this recipe and you can create separate PRs to add diagrams to the other ClickHouse recipes.
4. I believe `google-us-east4` is the correct region name. 
5. The Terraform code didn't show syntax highlighting. Kindly check.
6. Please create a card on the cookbook.rst landing page for this recipe.

wojcik-dorota commented:
> Hi, here are the review suggestions:
> 
> 1. Please use `version = "~> 3.9.0"` instead of `version = ">= 3.7"`
> 2. "This article shows by way of example..." phrase has been used already in the two other ClickHouse recipes. To avoid monotonous reading, please rephrase this sentence.
> 3. I missed to note in the previous two ClickHouse recipes but all of our cookbook recipes include Mermaid diagrams. Kindly include Mermaid diagrams for this recipe and you can create separate PRs to add diagrams to the other ClickHouse recipes.
> 4. I believe `google-us-east4` is the correct region name.
> 5. The Terraform code didn't show syntax highlighting. Kindly check.
> 6. Please create a card on the cookbook.rst landing page for this recipe.
> 7. The resultant database doesn't show any tables. For a validity of the story, it would be nice if there was a source table that shows up on the destination database using the integration.

@dewan-ahmed , thank you for the feedback :-) Addressed all your comments except the syntax highlighting - I'm not able to find what's wrong. Could you please help me identify the missing bit? All seems fine when I check.

dewan-ahmed commented:
Thanks @wojcik-dorota. I'm looking into why the syntax highlighting is not working. Tomorrow I'll propose a revised version of the Mermaid diagram.

mhoffm-aiven commented on docs/tools/terraform/reference/cookbook/postgres-clickhouse-integration-recipe.rst:
```diff
@@ -0,0 +1,183 @@
+    destination_service_name = aiven_clickhouse.clickhouse.service_name
+    clickhouse_postgresql_user_config {
+      databases {
+        database = aiven_pg_database. sensor_dims.database_name
```
note: extra whitespace after `aiven_pg_database`; this is probably root_cause of broken highlighting too

wojcik-dorota replied:
@mhoffm-aiven , thank you for spotting that. Indeed, with this fix, the syntax highlighting works fine :-)

dewan-ahmed commented:
Hi @wojcik-dorota @aris-aiven I'm proposing a new Mermaid diagram.

![image](https://user-images.githubusercontent.com/30604461/205981143-1d245c29-a0b6-4d3a-9c86-153ebf7e5dde.png)

The new diagram is much simpler, shows only one Aiven for ClickHouse block instead of two and points to the new database for the service integration.

If this looks OK, please use or you can keep the previous diagram - your call. One thing I'll point out is that I see the usage of **table** and **database** on this recipe whereas all three entities show up as databases on the Aiven console - sensor_dims, iot_measurements, and service_postgres-gcp-us_sensor_dims_public. If all of these are databases, please update accordingly (for example, line#9).

wojcik-dorota commented:
> 

Thank you, @dewan-ahmed :-) will follow up accordingly!

wojcik-dorota commented:
> 

Thank you, @dewan-ahmed :-) will follow up accordingly!

aris-aiven commented on docs/tools/terraform/reference/cookbook/postgres-clickhouse-integration-recipe.rst:
```diff
@@ -0,0 +1,180 @@
+Let's cook!
+-----------
+
+Imagine that you've been collecting IoT measurements from thousands of sensors and storing them in ClickHouse table ``iot_measurements``. Now, you'd like to enrich your metrics by adding the sensor's location measurement so that you can filter the metrics by the city name. The sensor's location data is available in the ``sensors_dim`` database in PostgreSQL.
```
@dewan-ahmed is correct that the introduction should read "database" instead of "table" for consistency and correctness

```suggestion
Imagine that you've been collecting IoT measurements from thousands of sensors and storing them in ClickHouse database ``iot_measurements``. Now, you'd like to enrich your metrics by adding the sensor's location measurement so that you can filter the metrics by the city name. The sensor's location data is available in the ``sensors_dim`` database in PostgreSQL.
```

aris-aiven reviewed (changes requested):
Left one final comment complementing other reviews.

aris-aiven reviewed (approved):
The content is comprehensible and the formatting is correct. Good to be merged in my opinion. Added some minor edits for clarity.

dewan-ahmed commented:
All good from my end @wojcik-dorota - thanks a lot for your patience with the review process.

A suggestion to rebase this branch (we're suggesting this for all open PRs) since https://github.com/aiven/devportal/pull/1620 just went in and we can fix any warning/error produced as a result. I can merge the PR right after.

dewan-ahmed commented:
Great! Thank you @wojcik-dorota & @aris-aiven
